cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -173,6 +204,19 @@ protected void run() {
         Thread.currentThread().interrupt();
       }
     }
+    // Persist records and messageId's after all ProcessMetricsThread's complete.
+    LinkedList<MetricValues> recordsCopy = new LinkedList<>();
```

Comment:
No need to copy, just pass the concurrent queue to the `persistRecords` (check the signature to take `Queue<MetricValues>`